### DISCUSSION TITLE
pop_screen callback patterns

### DISCUSSION THREAD

**Discussion by davetapley** [Textualize/textual#3402] [General]
I'm _hoping_ there's a nicer way to do this in Python... 

# Pattern 1

A `Screen` returns a value which needs to be assigned to a `reactive`.

```py
class SetupScreen(Screen[Server]):
    ...


class MyApp(App[None]):
    server: reactive[Server | None] = reactive(None)

    def update_server(self, server: Server):
        self.server = server

    def on_mount(self) -> None:
        self.push_screen(SetupScreen(), self.update_server)
```

Is there any way to do this without having to define and reference `update_server`?

# Pattern 2

A `Screen` which returns `None`.

```py
class SensorsScreen(Screen[None]):
   ...


class MyApp(App[None]):
    def clear_server(self, _: None):
        self.server = None

...

    self.push_screen(SensorsScreen(server.host, server.port), self.clear_server)
```

Is there any way to do this without needing ` _: None)` for the types to match?

**Comment by davep** [comment]
# Pattern 1

Short of `SetupScreen` having knowledge of the app that pushed it, I don't see a cleaner method. Most of the motivation behind the callback approach is to remove that level of knowledge so the screen can work about knowing the type of the app. Another approach, which is what I used to use before we added callbacks, is it post a message to the application from the screen before it pops itself; but then you need to define a message handler anyway.

What's the motivation for this? What are you trying to work around here? Simply removing the need for the callback function?

# Pattern 2

I don't believe so; `dismiss` is built around the idea that your screen will want to pass back *something* -- in the case here for example, perhaps a boolean value to state if the user indicated that they wanted the server to be cleared, or not (that's not to suggest that this is appropriate for this particular use -- I can't see what your `SensorsSceen` does of course -- but to describe the general use).

**Reply by davetapley** [reply]
# Pattern 1

> Simply removing the need for the callback function?

Yep, exactly this.
e.g. for a getter I could just pass `self.server`, but I'm not sure how (or if) you can pass equivalent of `self.server =` as a function? 🤔 

# Pattern 2

That makes sense. My use case is that I want a defined `server` for `SensorsScreen` to avoid `None` checks everywhere, but the user needs to be able to specify the host / port somewhere.

To do this I `watch_server` in the `App`, if it's `None` then I push `SetupScreen` to get info, else push `SensorsScreen` (which is the 'main app').

Then it's quite elegant to have a 'disconnect' button in `SensorsScreen` which does `self.dismiss`, but I need to `dismiss(None)` for the types to work.


... but I just realized passing back a value is fine, I just need a type with only one value
... and I know from my Haskell days that's the 'unit type'
... and lo and behold [the Wikipedia page for unit type](https://en.wikipedia.org/wiki/Unit_type) suggests `None` 😆
